### DISCUSSION TITLE
changing response inside test

### DISCUSSION THREAD

**Discussion by adammo94** [mswjs/msw#964] [Q&A]
I'm trying to do something pretty simpl.

I want to achieve something like this in one test:
1. test checks how many elements in list there are (msw) should be standard response (1 element)
2. test click on add form button, fills out the form and clicks on submit button
3. test checks how many elements in there are (msw), should be modified response (2 elements)

I saw server.use, however as far as I understand it it can only modify response once before the element is rendered.

**Comment by kettanaito** [comment]
Hey, @adammo94. 

I highly recommend creating a state in your mock and connecting it to the handlers. This way when your UI dispatches request that are meant to modify data (such as the submit button click), those requests get intercepted and modify the internal state. This will give you confidence that the request has been made and the UI state you receive is what your customers will see.

## (Recommended) Describe a data model

Consider using data modeling libraries like [@mswjs/data](https://github.com/mswjs/data). They will allow you to keep your data state manageable and support data modifications right from the handlers. 

```js
import { factory, primaryKey } from "@mswjs/data"
import { rest } from 'msw'
import { v4 } from 'uuid'

const data = factory({
  list: {
    id: primaryKey(v4())
  }
})

const handlers = [
  rest.get('/items', (req, res, ctx) => {
    return res(ctx.json(data.list.getAll()))
  }),

  rest.put('/items', (req, res, ctx) => {
    // Create a new item in the data whenever
    // the interface performs "PUT /items".
    data.list.create()
    return res(ctx.status(201), ctx.json(data.list.getAll()))
  })
]

const server = setupServer(...handlers)
// ...the rest of the setup (server.listen(), etc.)

it('my test', () => {
  render(<Component />)
  expect(items).toHaveLength(1)

  // Submit performs a "PUT /items" request
  // that creates a new item in the "data".
  click(submitButton)
  expect(items).toHaveLength(2)
})
```

## Introduce a state

If you are absolutely sure that modeling data is not acceptable in your case, introduce a state that you'd treat as the model above. I generally discourage you from this approach, as does nothing but put more weight on your shoulders.

```js
import { rest } from 'msw'

// Keep an array of items.
const items = []

const handlers = [
  rest.get('/items', (req, res, ctx) => {
    return res(ctx.json(items))
  }),

  rest.put('/items', (req, res, ctx) => {
    // Create a new item in the data whenever
    // the interface performs "PUT /items".
    items.push({ id: 'any-value' })
    return res(ctx.status(201), ctx.json(items))
  })
]

const server = setupServer(...handlers)
// ...the rest of the setup (server.listen(), etc.)

it('my test', () => {
  render(<Component />)
  expect(items).toHaveLength(1)

  // Submit performs a "PUT /items" request
  // that pushes a new item to the "items" array.
  click(submitButton)
  expect(items).toHaveLength(2)
})
```

Let me know if these suggestions help. 

---

> I saw server.use, however as far as I understand it it can only modify response once before the element is rendered.

No. `server.use` appends request handlers that take effect _immediately_ after they are appended. You can utilize that to change the response to an endpoint anywhere in your test, including amid the test. It has no connection over the component's rendering.

**Reply by joeprivettsmart** [reply]
If you have already added a request handler using `server.use`, how do you update that handler to return a different response mid-way through a test?

**Reply by berTrindade** [reply]
>> // Keep an array of items.
>> const items = []


@kettanaito If I'll be using the same state for all handlers, shouldn't it be a deep copy? E.g:

In this example below, I'm using `clone` fn from Ramda JS. And `getNodeRoleListResponse` is mocked data. So the original mocked data object isn't mutated.

```js
const makeNodeRolesList = () => {
  let listData = clone(getNodeRoleListResponse)
  return {
    getListData: () => listData,
    resetListData: () => {
      listData = clone(getNodeRoleListResponse)
    },
    updateRoleListData: data => {
      for (const index in listData) {
        if (listData[index].id === data.id) {
          listData[index] = data
          break
        }
      }
    },
    addListData: data => (listData = [...listData, data]),
    getDefaultData: () => clone(getNodeRoleListResponse),
  }
}
```

**Reply by kettanaito** [reply]
@joeprivettsmart, you "update" that already added handler by adding another handler via `server.use()` that captures the same resource. We currently don't maintain any handler identity, and the only way to affect the network at any point in time is to add a request handler. 

```js
it('one test', () => {
  server.use(rest.get('/resource', firstResolver))

  // some test actions...

  server.use(rest.get('/resource', secondResolver))

  // other test actions...
})
```

@berTrindade, I believe the suggestion was to introduce a _mutable_ state that can be altered and read by different request handlers. If your operations are pure, your changes won't transition between different handlers so you will inevitably stumble upon stale data. You may think of the `items` as a very primitive database. Your database operations aren't pure, so I was going with the mutability in this case as well. 

There's a certain degree of danger when introducing mutability but in this case, I think, it should be fine. One way to manage it is scoping the `items` to the particular test suite, which was precisely the example. It'd be interesting to experiment with a more friendly APIs to spawn and manage contexts in MSW in the future.

**Reply by berTrindade** [reply]
Nice! Thank you for the explanation, @kettanaito!
